### PR TITLE
Define rvalues and lvalues

### PR DESCRIPTION
**Pull request by andralex**
Did I miss something?

### CODE REVIEW

File: spec/expression.dd [hunk 1]
```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
...
+    $(LI `null`, `true`, `false`, `$`, `this`, `super`;)
+    $(LI literals: integral, floating point, character, string, array, associative array, function;)
+    $(LI `enum` values;)
+    $(LI the result of the following expressions:)
+    $(UL
+        $(LI functions that return by value, whether explicitly invoked or as the result of lowering
+        user-defined operators;)
+        $(LI the ternary operator `?:` when the two choices have different types, or at least one is an 
+        rvalue;)
+        $(LI built-in binary operators `||`, `&&`, `|`, `^`, `&`, `==`, `!=`, `is`, `!is`, `<`, `<=`, `>`,
```
Comment (thewilsonator) [inline spec/expression.dd]: I would reword that as the binary operators  [...] acting on builtin types.

File: spec/expression.dd [hunk 2]
```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
...
+        $(LI `is` expressions;)
+        $(LI the special keywords `__FILE__`, `__FILE_FULL_PATH__`, `__MODULE__`, `__LINE__`,
+        `__FUNCTION__`, `__PRETTY_FUNCTION__`, `__DATE__`, `__TIME__`, `__TIMESTAMP__`, `__VENDOR__`,
+        `__VERSION__`.)))
+
+$(DDOC_ANCHOR define-lvalue)$(P $(B Definition) ($(DOUBLEQUOTE Lvalue)): The following expressions, and no
+others, are called lvalue expressions or lvalues:)
+$(OL
+$(LI variable names;)
+$(LI function names;)
```
Comment (thewilsonator) [inline spec/expression.dd]: s/names/symbols

File: spec/expression.dd [hunk 3]
```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
...
+$(LI the result of the $(I DotIdentifier) grammatical construct `.` when the rightmost side of the dot
+is the name of a field (whether direct or `static`);)
```
Comment (thewilsonator) [inline spec/expression.dd]: ... and not an enum e.g. `struct a { enum b = 1; }`

File: spec/expression.dd [hunk 4]
```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
...
+$(LI the result of the following expressions:
+$(UL
+$(LI built-in unary operators `+`, `*`, `++` (prefix only), `--` (prefix only);)
```
Comment (thewilsonator) [inline spec/expression.dd]: Is not so sure about '+' (I get why, but I'm not sure the compiler actually treats it that way)

File: spec/expression.dd [hunk 5]
```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
...
+$(LI built-in binary operators `[]`, `=`, `+=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `~=`,
```
Comment (thewilsonator) [inline spec/expression.dd]: `[]` isn't a binary operator.

File: spec/expression.dd [hunk 6]
```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
...
+`<<=`, `>>=`, `>>>=`, `^^=`;)
+$(LI the ternary operator `?:` when the two choices have identical types and are lvalues;)
```
Comment (thewilsonator) [inline spec/expression.dd]: does `alias this` play any tricksies here?

File: spec/expression.dd (hunk 1 above)
Comment (thewilsonator) [reply spec/expression.dd]: fair enough.

File: spec/expression.dd (hunk 2 above)
Comment (andralex) [reply spec/expression.dd]: I see the grammar (same file, below) uses "identifier". I'll rephrase and wait for Walter.

File: spec/expression.dd (hunk 3 above)
Comment (andralex) [reply spec/expression.dd]: are enums fields? I think they're just enums that happen to be nested within the aggregate.

File: spec/expression.dd (hunk 4 above)
Comment (andralex) [reply spec/expression.dd]: I checked. See e.g. https://run.dlang.io/is/mPBBv1. But I realize the operand of `+` must be an lvalue, so I'll amend.

File: spec/expression.dd (hunk 6 above)
Comment (andralex) [reply spec/expression.dd]: Hrm, it actually does and in a not-super-consistent manner. Will update.

File: spec/expression.dd (hunk 3 above)
Comment (thewilsonator) [reply spec/expression.dd]: Oh, I missed the bit about fields.

**Comment by andralex** [comment]
I've redone the definitions, focusing on lvalues and defining rvalues by complement. Otherwise things get very tedious, e.g. "lvalues are the ternary operator if this and that and the other; rvalues are the ternary operator if not this or not that or not the other" etc.

**Comment by rainers** [comment]
You can grep https://github.com/dlang/dmd/blob/master/src/dmd/expression.d for `bool isLvalue()` to get the dirty details. Some have a comment to explain the special cases.

File: spec/expression.dd [hunk 7]
```diff
@@ -21,13 +21,44 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
-to `return`. This is because the parent of `x + f()` has function literal type, not expression type.
-
+to `return`. This is because the parent of `x + f()` has function literal type, not expression type.)
+
+$(DDOC_ANCHOR define-lvalue)$(P $(B Definition) ($(DOUBLEQUOTE Lvalue)): The following expressions, and no
+others, are called lvalue expressions or lvalues:)
+$(OL
+$(LI `this` inside `struct` and `union` member functions;)
+$(LI the result of the $(I DotIdentifier) grammatical construct `.` (left side may be missing) when the
+rightmost side of the dot is the name of a field (whether direct or `static`), variable, or function;)
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply spec/expression.dd]: Isn't the leftmost side supposed to be an lvalue?

File: spec/expression.dd [hunk 8]
```diff
@@ -21,13 +21,44 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
+$(LI the result of the following expressions:
+$(UL
+$(LI built-in unary operators `+` (when applied to an lvalue), `*`, `++` (prefix only), `--` (prefix only);)
```
Comment (tgehr) [inline spec/expression.dd]: Unary `+` indeed preserves lvalues, but I'm not convinced that it should.

File: spec/expression.dd [hunk 9]
```diff
@@ -21,13 +21,44 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
+circumstances:)
+$(OL 
+    $(LI $(I e$(SUBSCRIPT 1)) and $(I e$(SUBSCRIPT 1)) are lvalues of the same type; OR)
+    $(LI One of $(I e$(SUBSCRIPT 1)) and $(I e$(SUBSCRIPT 2)) is an lvalue of type `T` and the other has
+    and `alias this` converting it to `ref T`.))
+$(LI `mixin` expressions if and only if the compilation of the expression resulting from compiling
+the argument(s) to `mixin` is an lvalue;)
+$(LI `cast(U)` expressions applied to lvalues of type `T` when `T*` is implicitly convertible to `U*`;)
+$(LI `cast()` and `cast(`$(I qualifier list)`)` when applied to an lvalue.)
+)))
```
Comment (tgehr) [inline spec/expression.dd]: Missing:
- Identifiers referencing local variables.
- Results of `ref` function calls.

File: spec/expression.dd [hunk 10]
```diff
@@ -21,13 +21,44 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
+$(P Note: Although the definition is straightforward, a few subtleties exist related to function literals. In the
 statement `return (() => x + f())() * g();`, the full expression of `f()` is `x + f()`, not the expression passed
-to `return`. This is because the parent of `x + f()` has function literal type, not expression type.
-
+to `return`. This is because the parent of `x + f()` has function literal type, not expression type.)
+
+$(DDOC_ANCHOR define-lvalue)$(P $(B Definition) ($(DOUBLEQUOTE Lvalue)): The following expressions, and no
+others, are called lvalue expressions or lvalues:)
+$(OL
+$(LI `this` inside `struct` and `union` member functions;)
```
[the remark this answers is not in the document]
Comment (andralex) [reply spec/expression.dd]: @RazvanN7 this doesn't compile:

```d
class C { void f() { return fun(this); } }
void fun(ref C) {}
```

File: spec/expression.dd (hunk 7 above)
[the remark this answers is not in the document]
Comment (andralex) [reply spec/expression.dd]: >Is it still an rvalue if the rightmost side is a function? Maybe you meant ref/property function?

I think that's covered by function calls.

>Isn't the leftmost side supposed to be an lvalue?

No, e.g. this works:

```d
struct S { int field; }
void fun(ref int) {}
void gun() { fun(S().field); }
```

File: spec/expression.dd (hunk 8 above)
Comment (andralex) [reply spec/expression.dd]: @tgehr yah, documenting existing stuff instead of aspiring to perfection

Comment (andralex) [reply spec/expression.dd]: On second thought, "+ is a no-op" is easier to explain and understand than "+ is almost a no-op, you see, it copies the thing into an rvalue".

File: spec/expression.dd (hunk 7 above)
[the remark this answers is not in the document]
Comment (RazvanN7) [reply spec/expression.dd]: You mean this is also fine? 

```d
struct S { int field; }

void main()
{
    int *p = &S().field;   
}
```

Then we should close https://issues.dlang.org/show_bug.cgi?id=11920

File: spec/expression.dd (hunk 9 above)
Comment (andralex) [reply spec/expression.dd]: Thanks! Those got removed during some text reshuffling.

File: spec/expression.dd (hunk 7 above)
[the remark this answers is not in the document]
Comment (andralex) [reply spec/expression.dd]: @RazvanN7 that's not fine because the reference escapes and outlives the temporary expression.

[the remark this answers is not in the document]
Comment (andralex) [reply spec/expression.dd]: I was wrong, the case of member function calls wasn't covered. Fixed now, thanks.

**Comment by andralex** [comment]
>You can grep https://github.com/dlang/dmd/blob/master/src/dmd/expression.d for bool isLvalue() to get the dirty details. Some have a comment to explain the special cases.

cc @RazvanN7 :)

**Comment by andralex** [comment]
Updated to account for all feedback.

File: spec/expression.dd (hunk 8 above)
Comment (tgehr) [reply spec/expression.dd]: No, this makes no sense. Note that soon, + will perform integer promotion, and then promoted lvalues will most likely be rvalues. Much easier to just deprecate using +x as an lvalue completely, as otherwise either your explanation or your justification of the behavior is bound to become a lot more complicated. There is no need to innovate with unary +.

File: spec/expression.dd [hunk 11]
```diff
@@ -21,13 +21,47 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
+the argument(s) to `mixin` is an lvalue;)
+$(LI `cast(U)` expressions applied to lvalues of type `T` when `T*` is implicitly convertible to `U*`;)
+$(LI `cast()` and `cast(`$(I qualifier list)`)` when applied to an lvalue.)
+)))
+
+$(DDOC_ANCHOR define-rvalue)$(P $(B Definition) ($(DOUBLEQUOTE Rvalue)): Expressions that are not
+lvalues are rvalues.)
+
+$(P Note: Rvalues include all literals, special value keywords such as `__FILE__` and `__LINE__`,
+`enum` values, and the result of expressions not mentioned in the definition of lvalues above.)
```
Comment (WalterBright) [inline spec/expression.dd]: s/mentioned in the definition of/defined as/

File: spec/expression.dd [hunk 12]
```diff
@@ -21,13 +21,47 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
 statement `return (() => x + f())() * g();`, the full expression of `f()` is `x + f()`, not the expression passed
-to `return`. This is because the parent of `x + f()` has function literal type, not expression type.
-
+to `return`. This is because the parent of `x + f()` has function literal type, not expression type.)
+
+$(DDOC_ANCHOR define-lvalue)$(P $(B Definition) ($(DOUBLEQUOTE Lvalue)): The following expressions, and no
+others, are called lvalue expressions or lvalues:)
+$(OL
+$(LI `this` inside `struct` and `union` member functions;)
+$(LI a simple symbol or the result of the $(I DotIdentifier) grammatical construct `.` (left side may be
```
Comment (WalterBright) [inline spec/expression.dd]: `simple symbol` is not defined. It's erroneous anyway as a "simple symbol" that is a manifest constant is not an lvalue. I suggest replacing it with `variable`.

File: spec/expression.dd (hunk 11 above)
Comment (andralex) [reply spec/expression.dd]: done

**Comment by andralex** [comment]
@WalterBright OK thanks pushed

File: spec/expression.dd [hunk 13]
```diff
@@ -21,13 +21,47 @@ full expression of $(I expr$(SUBSCRIPT 1)). Otherwise, $(I expr) is its own full
...
+$(LI the result of the following expressions:
+$(UL
+$(LI built-in unary operators `+` (when applied to an lvalue), `*`, `++` (prefix only), `--` (prefix only);)
+$(LI built-in indexing operator `[]` (but not the slicing operator);)
+$(LI built-in assignment binary operators, i.e. `=`, `+=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `~=`,
+`<<=`, `>>=`, `>>>=`, and `^^=`;)
+$(LI the ternary operator $(I e) `?` $(I e$(SUBSCRIPT 1)) `:` $(I e$(SUBSCRIPT 2)) under the following
+circumstances:)
+$(OL
+    $(LI $(I e$(SUBSCRIPT 1)) and $(I e$(SUBSCRIPT 1)) are lvalues of the same type; OR)
```
Comment (TurkeyMan) [inline spec/expression.dd]: `1` and `2`??

Comment (andralex) [reply spec/expression.dd]: @TurkeyMan fixed straight in master, thanks
